pageserver/src/walingest.rs

```diff
@@ -1419,13 +1419,50 @@ impl WalIngest {
                 forknum: VISIBILITYMAP_FORKNUM,
             };
 
-            let mut vm_page_no = rec.blkno / pg_constants::VM_HEAPBLOCKS_PER_PAGE;
-            if rec.blkno % pg_constants::VM_HEAPBLOCKS_PER_PAGE != 0 {
-                // Tail of last remaining vm page has to be zeroed.
-                // We are not precise here and instead of digging in VM bitmap format just clear the whole page.
-                modification.put_rel_page_image_zero(rel, vm_page_no)?;
+            //
+            // Code copied from PostgreSQL `visibilitymap_prepare_truncate` function in `visibilitymap.c`
+            //
+
+            // last remaining block, byte, and bit
+            let mut vm_page_no = rec.blkno / (pg_constants::VM_HEAPBLOCKS_PER_PAGE as u32);
+            let trunc_byte = rec.blkno as usize % pg_constants::VM_HEAPBLOCKS_PER_PAGE
+                / pg_constants::VM_HEAPBLOCKS_PER_BYTE;
+            let trunc_offset = rec.blkno as usize % pg_constants::VM_HEAPBLOCKS_PER_PAGE
+                * pg_constants::VM_BITS_PER_HEAPBLOCK;
+
+            // Unless the new size is exactly at a visibility map page boundary, the
+            // tail bits in the last remaining map page, representing truncated heap
+            // blocks, need to be cleared. This is not only tidy, but also necessary
+            // because we don't get a chance to clear the bits if the heap is extended
+            // again.
+            if trunc_byte != 0 || trunc_offset != 0 {
+                let old_content = modification
+                    .tline
+                    .get_rel_page_at_lsn(rel, vm_page_no, Version::Modified(modification), ctx)
+                    .await?;
```

Comment:
The corresponding postgres code has this check:

```
		mapBuffer = vm_readbuf(rel, truncBlock, false);
		if (!BufferIsValid(mapBuffer))
		{
			/* nothing to do, the file was already smaller */
			return InvalidBlockNumber;
		}
```

We are missing that here, the `get_rel_page_at_lsn` call could fail if the VM is already smaller.

Reply:
Let's also add a test case for that scenario.

Reply:
I have spent a lot of time, but failed to create test reproducing this error (invalid page error).
To reproduce the problem we need to satisfy the following conditions:
1. Produce zero VM page. It can be done in several ways. For example we can some how detect the moment when VM page is zero extended but not yet initialised and create branch at this LSN. Or truncate table and perform vacuum, which cause Postgres to generate HRAP_TRUNCATE WAL record, which now (without this PR) cause writing of  zero page. 
2. Then we should mark this page as all-visible. It can be easily done by running vacuum. But vacuum will also update VM, replacing zero page with some initialized page. 
3. Then we need to force Postgres to clear visibility flag. It can be done by updating the record or inserting new record on the page. 
4. Finally we need to make Postgres read VM page from PS. It can be done by flushing shared buffers, for example using `select clear_buffer_cache();` from neon_utils. But throwing away dirty VM page cause savoimgit on PS, overwritting zero page.

So there are at least two challenges: how to make page as all visible but not save it to PS. And how to force loading of VM page without saving of dirty VM page to PS.

But the problem is really happen sometimes (rarely).
So I think that this PR as well as #9287 should be merged without test reproducing the issue.

Reply:
You can reach the ingestion codepath pretty easily:
```
create table foo (i int) with (autovacuum_enabled=off, fillfactor=10);
insert into foo select g from generate_series(1, 100000) g;
delete from foo where i > 1;
vacuum foo;
```

When the table is extended by the INSERT, it does not extend the VM. But the VACUUM will emit an SMGR_TRUNCATE_VM record for it anyway.

To then exercise the replay part, perform an index only scan I guess.

Reply:
Obviously I tried something like this.
This is the latest version of the test (before I have tried much more combinations of delete, update, vacuum ...).
Also I tried to do it explicitly and check in the WAL which WAL records are generated.
As I noticed above, the main problem that in all cases at 2) and 4) FPI of VM page is added to the WAL.
```
import pytest
from fixtures.neon_fixtures import NeonEnv


#
# Test that VM is properly truncated
#
def test_vm_truncate(neon_simple_env: NeonEnv):
    env = neon_simple_env

    n = 100
    endpoint = env.endpoints.create_start("main")

    con = endpoint.connect()
    cur = con.cursor()

    cur.execute("CREATE EXTENSION neon_test_utils")
    cur.execute("create table t(pk integer primary key, count integer default 0, filler text default repeat('?',200))")
    cur.execute(f"insert into t (pk) values (generate_series(1,{n}))")
    cur.execute("vacuum t")
    cur.execute("delete from t where pk > 1")
    cur.execute("vacuum t")

    cur.execute("update t set count=count+1")
    endpoint.clear_shared_buffers(cursor=cur)

    cur.execute("set enable_seqscan=off")
    cur.execute("select count(*) from t")
    assert cur.fetchall()[0] == (1,)
```

Reply:
Hmm, so VACUUM itself always creates a full-page image of the VM page? Then I don't understand how you can ever end up in a situation where you'd try to truncate a VM page that doesn't exist. Only VACUUM truncates the VM, but before it gets that far, it would always try to set the bits first, extending the VM, right?